dropshot/src/server.rs

```diff
@@ -154,6 +154,16 @@ impl<C: ServerContext> HttpServerStarter<C> {
         let server = builder.serve(make_service);
         info!(app_state.log, "listening");
 
+        if let Err(e) = crate::register_probes() {
```

Comment:
What happens if this is called multiple times in the same process?  What if it's concurrent?

Reply:
We deduplicate providers based on the **user-land address of the DOF structure**:

https://github.com/illumos/illumos-gate/blob/master/usr/src/uts/common/dtrace/dtrace.c#L15139-L15146

This made sense when we thought the DOF would be in some static, ro section, and that registration was just ioct-ing it downstairs. Because we're generating the DOF this isn't the case.

So... yeah. Unless I'm mistaken, this kind of sucks.

Perhaps if the `set_dtrace_probes` section were writable we could `cas` the first record to write a sentinel value and make sure no one sees those same records.

Reply:
Let me check my understanding, @ahl. Since we're reading some data from our special probes section into a buffer in memory, and the address of _that_ buffer is the value of the `dofhp_addr` field that DTrace uses to de-duplicate probes, it's possible that the same probe can be re-registered. Is that right?

That does sound pretty problematic. Some logic in USDT to ensure we only register probes once does sound important. Either a flag in each section, or a static variable in the library with the names of the registered probes sound like OK options. I'd have to think more about tradeoffs here.

Reply:
Just a note on the deduplication. This will be fixed in the next version of `usdt`. I'll wait for that to be merged, then update the PR accordingly.

Reply:
Done in b52dc29